Add Gas Estimation Guide (#7319)
ethereum/web3.js: docs/docs/guides/04_transactions/gas-and-fees.md

In addition to the calculated gas cost, an Ethereum transaction can specify an optional priority fee, which is an additional fee that is paid directly to the operator of the node that executes the transaction. Priority fees are intended to incentive node operators to execute transactions. A priority fee is specified as a value in addition to the base fee, which means that the total cost of the priority fee is always a factor of the amount of gas required to execute a transaction.

> With the above in mind, the total cost of fees associated with a transaction are calculated as: _units of gas used \* (base fee + priority fee)_.

Reviewer: ```suggestion
With the above in mind, the total cost of fees associated with a transaction is calculated as follows:

Total cost of fees = _units of gas used \* (base fee + priority fee)_
```
suggestion - i think writing it out on a new line and as a full math equation can make it easier for the user to read and stands out a bit more

## Estimating Gas

The Ethereum JSON-RPC specifies the [`eth_estimateGas` method](https://ethereum.org/en/developers/docs/apis/json-rpc/#eth_estimategas), which accepts a transaction and returns an estimate of the amount of gas required to execute that transaction. The transaction will not be executed and added to the blockchain. The estimate may be different (typically more) than the amount of gas actually used by the transaction for a variety of reasons, including EVM mechanics, node performance, and changes to the state of a smart contract. To invoke the `eth_estimateGas` RPC method, use the [`Web3Eth.estimateGas` method](/api/web3-eth/class/Web3Eth#estimateGas) and provide the [`Transaction`](/api/web3-types/interface/Transaction) for which to estimate gas.

> Web3.js transactions may specifying a gas limit (the maximum amount of gas they are able to consume) by providing the [`Transaction.gas` property](/api/web3/namespace/types#gas). If the specified gas limit is less than the actual amount of gas required to execute the transaction, the transaction will consume an amount of gas equal to the gas limit, which is not refunded, before failing and reverting any state changes made by the transaction.

Reviewer 2: ```suggestion
Web3.js transactions may specify a gas limit (the maximum amount of gas they are able to consume) by providing the [`Transaction.gas` property](/api/web3/namespace/types#gas). If the specified gas limit is less than the actual amount of gas required to execute the transaction, the transaction will consume an amount of gas equal to the gas limit, which is not refunded, before failing and reverting any state changes made by the transaction.
```

transaction.maxPriorityFeePerGas = feeData.maxPriorityFeePerGas;
```

## Generating Access Lists

An access list specifies the addresses and storage keys that a transaction plans to access. Specifying these elements in advance makes a transaction's gas costs more predictable.

The Ethereum JSON-RPC specifies the [`eth_createAccessList` method](https://github.com/ethereum/execution-apis/blob/4140e528360fea53c34a766d86a000c6c039100e/src/eth/execute.yaml#L54-L97), which accepts a transaction and returns an object that lists the addresses and storage keys that the transaction will access, as well as the approximate gas cost for the transaction if the access list is included. The transaction will not be executed and added to the blockchain. To invoke the `eth_createAccessList` RPC method, use the [`Web3Eth.createAccessList` method](/api/web3-eth/function/createAccessList) and provide the [`TransactionForAccessList`](/api/web3-types/interface/TransactionForAccessList) for which to generate the access list.

Web3.js transactions may specify an access list by providing the [`Transaction.accessList` property](/api/web3/namespace/types#accessList).

> ```ts
> const transaction: TransactionForAccessList = {
> 	from: '<SENDER ADDRESS>',
> 	to: '<RECEIVER ADDRESS>',
> 	value: web3.utils.ethUnitMap.ether,
> };
> 
> const accessListResult: AccessListResult = await web3.eth.createAccessList(transaction);
> transaction.accessList = accessListResult.accessList;
> transaction.gas = accessListResult.gasUsed;
> ```

Reviewer 2: I’m not sure about the rest of the team’s opinion, but from my experience, mutating data is generally not a good practice and often leads to unexpected outcomes. The example code would be clearer and more reliable if written as suggested below.
 However, feel free to disregard this suggestion if it doesn’t align with the general style of examples in the documentation.

```suggestion
```ts
const transactionDraft: TransactionForAccessList = {
	from: '<SENDER ADDRESS>',
	to: '<RECEIVER ADDRESS>',
	value: web3.utils.ethUnitMap.ether,
};

const accessListResult: AccessListResult = await web3.eth.createAccessList(transactionDraft);

const transaction: TransactionForAccessList = {
	...transactionDraft,
	accessList: accessListResult.accessList,
	gas: accessListResult.gasUsed,
};
```
```
